corehq/apps/data_analytics/models.py

```diff
@@ -157,3 +157,10 @@ def export_row(self, past_months):
             experienced_threshold=self.experienced_threshold or DEFAULT_EXPERIENCED_THRESHOLD,
             performance_threshold=self.performance_threshold or DEFAULT_PERFORMANCE_THRESHOLD,
         )
+
+
+class DomainStats(models.Model):
+    domain = models.TextField(unique=True, db_index=True)
+    stats = models.JSONField()
```

Comment:
Do all stats have the same data type? Could they be normalized out so this table would have a `name` and `value` field, and there would be one row for each stat name/value pair?

Denormalization can also be achieved if values have different data types, although it gets a bit more complicated.

Reply:
Ah sorry I should have highlighted the rationale for lumping into one json field. There are a quite a few [calculated properties](https://github.com/dimagi/commcare-hq/blob/dc4998b7810711c504259ca4f11348cbcb06936a/corehq/apps/domain/calculations.py#L354-L409), so it felt "easiest" to just lump them into one field, though we had only considered completely denormalizing data into one row, not a row for each domain/calculated property pair. That does sound interesting, and it would still be easy to collect all calculated properties for a domain 🤔

Reply:
I also didn't answer your question. Apart from some boolean fields (which could be represented by an integers) and the "cp_last_updated" field which could be replaced by the last_modified field, these are all integers

Reply:
Normalizing the data would make it MUCH easier to run aggregate queries across multiple domains, which seems like it may have some interesting use cases.

Reply:
@millerdev One thing I'd be concerned about is the performance of updating ~65 rows per domain every night, instead of a large value in a single row like using a combined JSONField would do. Even if we're only conditionally updating rows, and also saying that maybe 10 of the properties change any given day on an average domain, it seems like we'd still have to at least access all 65 properties for those domains to check whether they need updating. The task to update calculated properties currently takes ~2 hours to run when writing to ES, and I worry that would be even longer with one SQL row per property.

Some caveats though, are that I don't personally know how ES compares to SQL in terms of efficiency, so it's possible it would still be fine, and also I might missing some obvious optimization here.

Reply:
Ended up adding a dedicated column for every property as that seemed like the simplest approach.

Reply:
Also just noting that there are guaranteed to be columns that don't need to be calculated once initially set (like 300th form submission). It could be nice to build that into the task that ultimately calculates these values to first check if it needs to in the cases where applicable.